File: python/create_noise.py
import os
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import savemat
from tkinter import Tk, filedialog
import pyabf

# Set the correct working directory
wd = os.path.dirname(os.path.abspath(__file__))
os.chdir(wd)

# Prompt window to select the file
root = Tk()
root.withdraw()
fileName = filedialog.askopenfilename(title="Select Data File", filetypes=[("ABF files", "*.abf")])
root.destroy()

# Load the data
AS = np.load(fileName)
AS = AS - np.convolve(AS, np.ones(10000) / 10000, mode='valid')  # Remove slow data deviations

# Create the noise trace
fig, ax = plt.subplots()
ax.plot(AS)
ax.set(title="Select the region with only noise (left-click start, right-click end)")
noise_regions = plt.ginput(n=-1, timeout=-1)
plt.close()

ASnoise = np.zeros_like(AS)
for start, end in noise_regions:
    start, end = int(start), int(end)
    ASnoise[start:end] = AS[start:end]

ASnoise = ASnoise - np.convolve(ASnoise, np.ones(10000) / 10000, mode='valid')  # Remove slow data deviations

# Save the noise trace
noise_file_name = os.path.splitext(os.path.basename(fileName))[0] + '_noise.mat'
save_path = os.path.join(wd, 'data_to_test', noise_file_name)
savemat(save_path, {"ASnoise": ASnoise})

File: MATLAB/create_noise.m
%% create the noise file. 

% set the correct working directory (same directory this file is located in) 
wd = matlab.desktop.editor.getActive;
wd = fileparts(wd.Filename);
cd(wd);
% addpath 'helper_functions';
[fileName,filePath] = uigetfile('*.abf');

% load the data 
AS = abfload([filePath, fileName]);  % select the data you want to run
AS = AS - movmean(AS, 10000);  % take out slow data deviations

% create the noise trace. When each figure window is prompted, select the 
% region you believe contains only noise by left clicking on the start of
% the region then right click on the end of the region. Continue until you 
% have adequate length of noise trace then press enter to finish
noiseFileName = [fileName, '_noise.mat'];
[ASnoise] = MakeNoiseFromFile(AS);
ASnoise = ASnoise - movmean(ASnoise, 10000);% take out slow data deviations

%% save the noise trace created
save([wd, '/data_to_test/', noiseFileName], "ASnoise");